packages/runtime/container-runtime/src/opLifecycle/outbox.ts

```diff
@@ -90,25 +106,34 @@ export class Outbox {
 			0x58d /* Reference sequence numbers from both batches must be in sync */,
 		);
 
+		const currentClientSequenceNumber = this.params.getProcessedClientSequenceNumber();
+
 		if (
 			(mainBatchReference === undefined ||
 				mainBatchReference === message.referenceSequenceNumber) &&
 			(attachFlowBatchReference === undefined ||
-				attachFlowBatchReference === message.referenceSequenceNumber)
```

Comment:
This IF very hard to read.
Some suggestions:
- instead of exposing reference seq# & clientSeq# as two properties, expose a single property that contains both of them. The Ideally it should be an object with two props, but if it complicates some operations, we can also go with a string in format of "seq#.clientSeq#". 
- With your changes fully in, I think we can retire "attachment batch" concept. While it might give us some latency improvement, I think complexity of it is not worth it.
- (partially answering your question addressed to me in other comment): Should we add message.referenceClientSequenceNumber instead of exposing params.getProcessedClientSequenceNumber()? referenceClientSequenceNumber would be valid only for outbound ops that are "in transition", i.e. only while we are processing them before they are sent to service - this property would not make it to loader.

And somewhat related:
-  I think message.referenceSequenceNumber is the same as IDeltaManager.lastSequenceNumber, which is same as IDeltaManager.lastMessage.sequenceNumber. While you are changing concepts (IDeltaManager.lastMessage.clientSequenceNumber would have different meaning from currentClientSequenceNumber used here), if my assumption is correct, I'd rather see DeltaManagerProxy changed (or rather new proxy created at runtime level) to expose consistent view for runtime. I.e. such that IDeltaManager.lastMessage.clientSequenceNumber to be same as currentClientSequenceNumber here (i.e. being updated on every op processed, and be index into current batch, not number of grouped ops client sent on current connection).